### PR TITLE
Tracker: an easy way to configure Wi-Fi/GNSS/FuelGauge as a wakeup source

### PR DESCRIPTION
**Pull request by XuGuohui**
Currently if we want to utilize the Wi-Fi/GNSS/FuelGauge to wake up device from sleep mode, we need to attach an interrupt to the INT pin of these peripherals followed by specifying the `IOE_INT`, which is the INT pin of the IO expander, as the wakeup source. For instance:

```
void handler() {
    // Do nothing
}

void sleepNow() {
    System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::STOP).gpio(IOE_INT, FALLING));
}

void setup() {
    attachInterrupt(GPS_INT, handler, FALLING);
}
```

This PR will simplify the code to configure the INT pins of these peripherals as a wakeup source directly:

```
void sleepNow() {
    System.sleep(SystemSleepConfiguration().mode(SystemSleepMode::STOP).gpio(GPS_INT, FALLING));
}
```

### Example App

Test application added within this PR.

### References

N/A

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

**Review by avtolstoy — APPROVED** [review]

File: hal/src/nRF52840/sleep_hal.cpp [hunk 1]
```diff
@@ -48,6 +48,11 @@
 #endif
 #include "spark_wiring_vector.h"
 
+#if HAL_PLATFORM_IO_EXTENSION && MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
```
[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/nRF52840/sleep_hal.cpp:51]: We have to sprinkling the `HAL_PLATFORM_IO_EXTENSION` and `HAL_PLATFORM_MCP23S17` anyway. And potentially in the future the bootloader may involve sleep function as well.

File: system/src/system_power_manager.h [hunk 2]
```diff
@@ -31,6 +31,7 @@ class PowerManager {
 
   void init();
   void sleep(bool s = true);
```
[the remark this answers is not in the document]
Comment (XuGuohui) [reply system/src/system_power_manager.h:33]: Will do.

File: hal/src/tracker/mcp23s17.cpp [hunk 3]
```diff
@@ -311,6 +311,37 @@ int Mcp23s17::detachPinInterrupt(uint8_t port, uint8_t pin, bool verify) {
...
+
+int Mcp23s17::interruptsRestore(uint8_t intStatus[2]) {
+    Mcp23s17Lock lock();
+    CHECK_TRUE(initialized_, SYSTEM_ERROR_INVALID_STATE);
+    CHECK(readRegister(INTF_ADDR[0], &intStatus[0]));
+    CHECK(readRegister(INTF_ADDR[1], &intStatus[1]));
+    CHECK(writeRegister(INTCON_ADDR[0], intcon_[0]));
+    CHECK(writeRegister(INTCON_ADDR[1], intcon_[1]));
+    CHECK(writeRegister(DEFVAL_ADDR[0], defval_[0]));
+    CHECK(writeRegister(DEFVAL_ADDR[1], defval_[1]));
```
[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/tracker/mcp23s17.cpp:339]: Registers' values are cached using member variables like `defval_`. They are reset on initialization and updated every time a register is written.

File: hal/src/nRF52840/sleep_hal.cpp [hunk 4]
```diff
@@ -332,27 +374,88 @@ static void configGpioWakeupSource(const hal_wakeup_source_base_t* wakeupSources
...
+            if (nrf_gpiote_event_pin_get(i) != ioeNrfPin) {
+                nrf_gpiote_event_clear(gpioteEvents[i]);
+            }
+        }
         nrf_gpiote_event_clear(NRF_GPIOTE_EVENTS_PORT);
         NVIC_ClearPendingIRQ(GPIOTE_IRQn);
-
         nrf_gpiote_int_enable(gpioIntenSet);
+
+        NVIC_ClearPendingIRQ(GPIOTE_IRQn);
```
[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/nRF52840/sleep_hal.cpp:401]: It's a typo. Will remove it.

File: hal/src/nRF52840/sleep_hal.cpp [hunk 5]
```diff
@@ -523,9 +626,29 @@ static bool isWokenUpByGpio(const hal_wakeup_source_gpio_t* gpioWakeup) {
...
+    }
+#endif
+    else {
+        return false;
+    }
+#endif
+
     if (nrf_gpiote_event_is_set(NRF_GPIOTE_EVENTS_PORT)) {
         // PORT event
-        uint32_t nrfPin = NRF_GPIO_PIN_MAP(halPinMap[gpioWakeup->pin].gpio_port, halPinMap[gpioWakeup->pin].gpio_pin);
```
[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/nRF52840/sleep_hal.cpp:528]: Yup 👍 . That could simplify the code.

File: hal/src/tracker/mcp23s17.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/src/tracker/mcp23s17.cpp:339]: That's what I'm trying to say. The outlined registers (INTCONA/B and DEFVALA/B) are not cached anywhere. I saw that it's done for some others, but these are not cached anywhere.

[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/tracker/mcp23s17.cpp:339]: Those are const values that is defined in mcp23s17.h

[the remark this answers is not in the document]
Comment (avtolstoy) [reply hal/src/tracker/mcp23s17.cpp:339]: There is nothing defined in mcp23s17.h. The only place these member variables are modified in particular is `resetRegValue()`.

[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/tracker/mcp23s17.cpp:339]: For the `intcon_[]` and `defval_[]`, there is no way to update them outside of sleep code. So their value will be the same as reset value. If we're going to modify them in the future, we'll cache them as well, like what we've done to other config registers.

**Pull request merged**